### ISSUE TITLE
leJOS EV3 version install faliure

### ISSUE THREAD

**Issue by johnmeshulam** [ev3dev-lang-java/openjdk-ev3#50]
Hi,
I've recently tried to install the leJOS ev3 version of openJDK 9 and 10.
I downloaded the tar.gz files of release [0.4.5.1](https://github.com/ev3dev-lang-java/openjdk-ev3/releases/tag/v0.4.5.1-lejos) and tried installing in 2 ways on a newly formatted SD card:

1. 
What did I do? 
- Place the tar.gz file in the top directory of the SD, unzip lejosimage, boot the EV3.

What happened?

- The installation failed with the message 'Java is not installed'

2. 
What did I do? 
- Use an SD card with running leJOS that currently has a ejre-8 on it, delete the old jre, place the new tar.gz file, and boot. 

What happened? 

- Everything worked fine and java was installed, but the brick got stuck on the "Start Menu" stage (final stage of leJOS boot) and never started. I tried this multiple times.

I'd appreciate help solving this, if it is a problem on my side, or a fix if not

**Comment by JakubVanek** [comment]
Hello @johnmeshulam,

yes, this is a known issue and it is not a problem on your side. Sadly, the builds of OpenJDK provided by this project will not work on leJOS. I have looked at this project's releases and they indeed say `release for leJOS`, which is misleading. This is an error from the times when I thought that the builds might work there.

The reason is that these binaries are built on and for newer Debian systems (Jessie _might_ work, however Stretch is the earliest officially supported version) that have newer libraries (glibc) that are not compatible with the system core that leJOS uses (Linux kernel 2.6.x; newer glibc versions require at least version 3.2.x). Thus it is impossible to run these binaries on leJOS, as they will immediately fail to run with the message below (source [here][link], but it's necessary to click on `quote` button on the last but one post).
```
FATAL: kernel too old
```
[link]: https://lejos.sourceforge.io/forum/viewtopic.php?f=18&t=8690#p33113

So as far as this project goes, the only Java versions usable on leJOS are the ones that [Oracle has provided][oracle]. That way it's possible to install Java 7 easily and Java 8 with some hurdles (it is necessary to generate EJDK with a correct profile from the provided tarball, I think it's `compact3`).

[oracle]: https://www.oracle.com/java/technologies/javaseemeddedev3-downloads.html

If we enter the theoretical realm, it should be possible to build this OpenJDK in a similar way for leJOS. However, one would have to abandon Debian software repositories (at least the ones of the newer versions). Then there are two options:
 * First option is to build the necessary packages from sources using approximately the OpenEmbedded-Classic version that the LEGO-originated (and leJOS-reused) operating system is built from. It is somehow derived from the Ångström distribution, but the webpages of that project have changed in the meantime. The system also contains a string like "Rudolf 2011.2" and I think that I found somewhere a Texas Instruments branch for AM1808 (the chip used in the EV3) and similar chips.
 * Alternatively, it _might_ be possible to use the repositories of a Debian (or another distro) version old enough that the binaries built there would be runnable on leJOS. It seems that some debian packages in the leJOS SDCard repository come from Debian Wheezy, so that might be the compatible version.

I so far haven't gone down either of the roads. However if I had to choose today, I would try the second option first. I may take a look at it, but I don't want to give some guarantees that I wouldn't later fulfill.

Thanks for asking and best regards,

Jakub

**Comment by JakubVanek** [comment]
The second build option has an interesting twist:
* Cross build (x86->ARM) would require cross compilers that are missing on Debian Wheezy (and Emdebian is dead). Alternative would be to use Ubuntu Precise as that seems close enough and it has cross toolchains in repositories.
* Native build (ARM->ARM) would require a recent build of OpenJDK 10 or 11 that would work on this old distro and this architecture. This rules out reusing this project's build, because they require new glibc. Azul's Zulu provides aarch32sf builds only for JDK8, so that does not help us. It would therefore be necessary to build some of the previous versions (Azul's 8 -> our 9, our 9 -> our 10, our 10 -> our 11).
Using Ubuntu Precise with an external OpenJDK11 seems like the next step.

**Comment by johnmeshulam** [comment]
Thanks for all the help! @JakubVanek 

I'll definitely take a loot at building with the second option you wrote and see what I can do.

Ultimately my goal is getting lambda function to work properly on leJOS EV3.
I managed to get Java 8 on leJOS a while ago, but lambda functions have a solid 5 second delay at runtime every time they are called...

On EV3dev lambdas work fine, so I thought I'd start by getting a newer JDK on leJOS. 
Do you have any idea why this delay at runtime could be caused?

**Comment by JakubVanek** [comment]
Hi @johnmeshulam ,

I don't have experience with Java performance characteristics regarding lambdas, but I've found one issue on StackOverflow that reaches the conclusion that lambdas can be slow in some contexts: https://stackoverflow.com/a/34642063

Here's what I remember that might relate to this issue:
* Classloading takes a majority of the JVM startup time, at least on ev3dev with ev3dev-lang-java.
* We have also encountered that JShell (interactive java shell) which has to do some sort of runtime compilation was nearly unusable when running both the frontend-backend combo and backend only directly on the brick. This was however likely due to the memory constraints on the brick.

It may be that in later OpenJDK releases indeed the lambda expressions are optimized.

Best regards,

Jakub
